core/src/main/java/org/apache/iceberg/PositionDeletesTable.java

```diff
@@ -0,0 +1,396 @@
+                "Partition that position delete row belongs to"),
+            Types.NestedField.required(
+                POSITION_DELETE_TABLE_SPEC_ID,
+                "spec_id",
+                Types.IntegerType.get(),
+                "Spec ID of the file that the position delete row belongs to"),
+            Types.NestedField.required(
+                POSITION_DELETE_TABLE_FILE_PATH_ID,
+                "delete_file_path",
+                Types.StringType.get(),
+                "Path of the delete file that the position delete row belongs to"));
+
+    if (partitionType.fields().size() > 0) {
+      return result;
+    } else {
+      // avoid returning an empty struct, which is not always supported.
+      // instead, drop the partition field
+      return TypeUtil.selectNot(result, Sets.newHashSet(POSITION_DELETE_TABLE_PARTITION_FIELD_ID));
+    }
+  }
+
+  public static class PositionDeletesBatchScan
+      extends SnapshotScan<BatchScan, ScanTask, ScanTaskGroup<ScanTask>> implements BatchScan {
+
+    protected PositionDeletesBatchScan(TableOperations ops, Table table, Schema schema) {
+      super(ops, table, schema, new TableScanContext());
+    }
+
+    protected PositionDeletesBatchScan(
+        TableOperations ops, Table table, Schema schema, TableScanContext context) {
+      super(ops, table, schema, context);
+    }
+
+    @Override
+    protected PositionDeletesBatchScan newRefinedScan(
+        TableOperations newOps, Table newTable, Schema newSchema, TableScanContext newContext) {
+      return new PositionDeletesBatchScan(newOps, newTable, newSchema, newContext);
+    }
+
+    @Override
+    public CloseableIterable<ScanTaskGroup<ScanTask>> planTasks() {
+      return TableScanUtil.planTaskGroups(
+          planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+    }
+
+    protected List<String> scanColumns() {
+      return context().returnColumnStats() ? DELETE_SCAN_WITH_STATS_COLUMNS : DELETE_SCAN_COLUMNS;
+    }
+
+    @Override
+    protected CloseableIterable<ScanTask> doPlanFiles() {
+      Expression rowFilter = filter();
+      String schemaString = SchemaParser.toJson(tableSchema());
+
+      // prepare transformed partition specs and caches
+      Map<Integer, PartitionSpec> transformedSpecs =
+          table().specs().values().stream()
+              .map(spec -> transformSpec(tableSchema(), spec))
+              .collect(Collectors.toMap(PartitionSpec::specId, spec -> spec));
+
+      LoadingCache<Integer, ResidualEvaluator> residualCache =
+          Caffeine.newBuilder()
+              .build(
+                  specId -> {
+                    PartitionSpec spec = transformedSpecs.get(specId);
+                    Expression filter =
+                        context().ignoreResiduals() ? Expressions.alwaysTrue() : filter();
+                    return ResidualEvaluator.of(spec, filter, isCaseSensitive());
+                  });
+      LoadingCache<Integer, String> specStringCache =
```

Comment:
nit: Add an empty line between these blocks as they are relative big and independent?

Reply:
Not sure if its what you meant, but added line between all these caches.